.pre-commit-config.yaml

```diff
@@ -0,0 +1,10 @@
+repos:
+  - repo: https://github.com/psf/black
+    rev: 22.12.0
```

Comment:
What's the interaction between the `black` from pip and this revision-pegged black? I feel like we'd do one or the other, but this PR has both.

It doesn't seem crazy to just skip `pre-commit` altogether and just ship a `.git/hooks/pre-commit` like:
```
#!/bin/sh
black .
```

But I assume this is the tip of the spear of a whole list of pre-commits you envision :smile:

Reply:
so i think you're right about the redundancy - let me check and verify.

and yeah, we can add other pre-commit hooks, like bandit, for one.

Reply:
That's where we'll have a more interesting debate -- I am not (currently) in favor of adding gatekeeping checks as `pre-commits`. But we'll see.

Speaking of which, what happens if I don't install the pre-commit hook and write bad code? Should this PR also add a `ci.yaml` check that black returns 0?

Reply:
while possible, i think it's overkill - you'll get linter errors on CI and the response is 'please install the commit hooks'.

Reply:
You wouldn't necessarily get linter errors on CI (unless we add this `black` check). Black and the linter setup we have don't 100% align (c.f. all the changes in this PR).

This distinction of gatekeeping on commit vs CI is similar in my mind to "do I validate user input on the web client or the server?" -- I hope we'd agree that you NEED to validate on the server and MAY validate on the client if it improves UX.

If we don't run `black` as part of the CI, a lazy developer (or contributor that didn't pore over our docs and assumed a normal repo you can just walk up to and edit -- not a crazy thing) can contribute code that black would adjust for the next person. For example, reviewers would be unlikely to notice that there should be a newline between a docstring and the first method.

The bad outcome here of something squeezing through is so tiny it doesn't really matter. But I think the value might be more of a enforcement check on someone actually using our recommended developer setup in case they forgot. "Ah yeah, I didn't install my pre-commits"

Reply:
ok, added a black step